File: FloatingShortcutsPhone/src/main/java/net/geekstools/floatshort/PRO/Widget/NavAdapter/ConfiguredWidgetsAdapter.java
package net.geekstools.floatshort.PRO.Widget.NavAdapter;

import android.app.Activity;
import android.appwidget.AppWidgetHost;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProviderInfo;
import android.content.Context;
import android.graphics.drawable.GradientDrawable;
import android.graphics.drawable.LayerDrawable;
import android.util.DisplayMetrics;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import androidx.sqlite.db.SupportSQLiteDatabase;

import com.google.android.material.textfield.TextInputEditText;

import net.geeksempire.chat.vicinity.Util.RoomSqLiteDatabase.UserInformation.WidgetDataDAO;
import net.geeksempire.chat.vicinity.Util.RoomSqLiteDatabase.UserInformation.WidgetDataInterface;
import net.geekstools.floatshort.PRO.R;
import net.geekstools.floatshort.PRO.Util.Functions.FunctionsClass;
import net.geekstools.floatshort.PRO.Util.Functions.PublicVariable;
import net.geekstools.floatshort.PRO.Util.NavAdapter.NavDrawerItem;
import net.geekstools.floatshort.PRO.Widget.WidgetConfigurations;

import java.util.ArrayList;

public class ConfiguredWidgetsAdapter extends RecyclerView.Adapter<ConfiguredWidgetsAdapter.ViewHolder> {

    Activity activity;
    Context context;

    FunctionsClass functionsClass;

    ArrayList<NavDrawerItem> navDrawerItems;

    View view;
    ViewHolder viewHolder;

    AppWidgetManager appWidgetManager;
    AppWidgetHost appWidgetHost;

    public ConfiguredWidgetsAdapter(Activity activity, Context context, ArrayList<NavDrawerItem> navDrawerItems, AppWidgetManager appWidgetManager, AppWidgetHost appWidgetHost) {
        this.context = context;
        this.activity = activity;

        this.navDrawerItems = navDrawerItems;

        this.appWidgetManager = appWidgetManager;
        this.appWidgetHost = appWidgetHost;

        functionsClass = new FunctionsClass(context, activity);
    }

    @Override
    public ConfiguredWidgetsAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        view = LayoutInflater.from(context).inflate(R.layout.widget_configured_items, parent, false);
        viewHolder = new ViewHolder(view);
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(ViewHolder viewHolderBinder, final int position) {
        AppWidgetProviderInfo appWidgetProviderInfo = navDrawerItems.get(position).getAppWidgetProviderInfo();
        int appWidgetId = navDrawerItems.get(position).getAppWidgetId();

        WidgetConfigurations.createWidget(activity, viewHolder.widgetPreview,
                appWidgetManager, appWidgetHost,
                appWidgetProviderInfo, appWidgetId);

        viewHolder.widgetLabel.setText(navDrawerItems.get(position).getAddedWidgetRecovery() ? navDrawerItems.get(position).getWidgetLabel() + " " + "\uD83D\uDD04" : navDrawerItems.get(position).getWidgetLabel());
        viewHolder.widgetLabel.setTextColor(PublicVariable.themeLightDark ? context.getColor(R.color.dark) : context.getColor(R.color.light));

        LayerDrawable drawFloatTheWidget = (LayerDrawable) context.getDrawable(R.drawable.draw_open);
        GradientDrawable backFloatTheWidget = (GradientDrawable) drawFloatTheWidget.findDrawableByLayerId(R.id.backtemp);
        backFloatTheWidget.setTint(functionsClass.extractDominantColor(functionsClass.appIcon(navDrawerItems.get(position).getPackageName())));
        viewHolder.floatTheWidget.setImageDrawable(drawFloatTheWidget);

        viewHolder.floatTheWidget.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                functionsClass.runUnlimitedWidgetService(navDrawerItems.get(position).getAppWidgetId(), navDrawerItems.get(position).getWidgetLabel());
            }
        });
        viewHolder.floatTheWidget.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View view) {
                functionsClass.doVibrate(77);

                functionsClass.popupOptionWidget(context, view,
                        navDrawerItems.get(position).getPackageName(),
                        navDrawerItems.get(position).getAppWidgetId(),
                        navDrawerItems.get(position).getWidgetLabel(),
                        appWidgetProviderInfo.loadPreviewImage(context, DisplayMetrics.DENSITY_LOW) != null ? appWidgetProviderInfo.loadPreviewImage(context, DisplayMetrics.DENSITY_HIGH) : appWidgetProviderInfo.loadIcon(context, DisplayMetrics.DENSITY_HIGH),
                        navDrawerItems.get(position).getAddedWidgetRecovery());

                return false;
            }
        });

        viewHolder.widgetLabel.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView textView, int actionId, KeyEvent keyEvent) {
                if (actionId == EditorInfo.IME_ACTION_DONE) {
                    if (textView.length() > 0) {
                        new Thread(new Runnable() {
                            @Override
                            public void run() {
                                WidgetDataInterface widgetDataInterface = Room.databaseBuilder(context, WidgetDataInterface.class, PublicVariable.WIDGET_DATA_DATABASE_NAME)
                                        .fallbackToDestructiveMigration()
                                        .addCallback(new RoomDatabase.Callback() {
                                            @Override
                                            public void onCreate(@NonNull SupportSQLiteDatabase supportSQLiteDatabase) {
                                                super.onCreate(supportSQLiteDatabase);
                                            }

                                            @Override
                                            public void onOpen(@NonNull SupportSQLiteDatabase supportSQLiteDatabase) {
                                                super.onOpen(supportSQLiteDatabase);

                                            }
                                        })
                                        .build();
                                WidgetDataDAO widgetDataDAO = widgetDataInterface.initDataAccessObject();
                                widgetDataDAO.updateWidgetLabelByWidgetId(navDrawerItems.get(position).getAppWidgetId(), textView.getText().toString().replace("\uD83D\uDD04", ""));
                                widgetDataInterface.close();
                            }
                        }).start();
                    }

                }
                return true;
            }
        });
    }

    @Override
    public int getItemCount() {
        return navDrawerItems.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {

        RelativeLayout widgetItem, widgetPreview;
        ImageView floatTheWidget;
        TextInputEditText widgetLabel;

        public ViewHolder(View view) {
            super(view);
            widgetItem = (RelativeLayout) view.findViewById(R.id.widgetItem);
            widgetPreview = (RelativeLayout) view.findViewById(R.id.widgetPreview);
            widgetLabel = (TextInputEditText) view.findViewById(R.id.widgetLabel);
            floatTheWidget = (ImageView) view.findViewById(R.id.floatTheWidget);
        }
    }
}

File: FloatItPhone/src/main/java/net/geekstools/floatshort/PRO/Widgets/WidgetsAdapter/ConfiguredWidgetsAdapter.kt
/*
 * Copyright © 2020 By Geeks Empire.
 *
 * Created by Elias Fazel
 * Last modified 8/25/20 4:42 AM
 *
 * Licensed Under MIT License.
 * https://opensource.org/licenses/MIT
 */

package net.geekstools.floatshort.PRO.Widgets.WidgetsAdapter

import android.appwidget.AppWidgetHost
import android.appwidget.AppWidgetManager
import android.graphics.drawable.LayerDrawable
import android.util.DisplayMetrics
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import android.widget.ImageView
import android.widget.RelativeLayout
import androidx.recyclerview.widget.RecyclerView
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.sqlite.db.SupportSQLiteDatabase
import com.google.android.material.textfield.TextInputEditText
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import net.geekstools.floatshort.PRO.R
import net.geekstools.floatshort.PRO.SearchEngine.UI.SearchEngine
import net.geekstools.floatshort.PRO.Utils.AdapterItemsData.AdapterItems
import net.geekstools.floatshort.PRO.Utils.Functions.FunctionsClass
import net.geekstools.floatshort.PRO.Utils.Functions.PublicVariable
import net.geekstools.floatshort.PRO.Widgets.RoomDatabase.WidgetDataInterface
import net.geekstools.floatshort.PRO.Widgets.Utils.FunctionsClassWidgets
import net.geekstools.floatshort.PRO.Widgets.WidgetConfigurations
import java.util.*

class ConfiguredWidgetsAdapter(private val widgetConfigurationsActivity: WidgetConfigurations,
                               private val adapterItems: ArrayList<AdapterItems>, private val appWidgetManager: AppWidgetManager, private val appWidgetHost: AppWidgetHost) : RecyclerView.Adapter<ConfiguredWidgetsAdapter.ViewHolder>() {

    private val functionsClass: FunctionsClass = FunctionsClass(widgetConfigurationsActivity)
    private val functionsClassWidgets: FunctionsClassWidgets = FunctionsClassWidgets(widgetConfigurationsActivity)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {

        return ViewHolder(LayoutInflater.from(widgetConfigurationsActivity).inflate(R.layout.widget_configured_items, parent, false))
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {

        val appWidgetProviderInfo = adapterItems[position].appWidgetProviderInfo
        val appWidgetId = adapterItems[position].appWidgetId

        widgetConfigurationsActivity.createWidget(widgetConfigurationsActivity, viewHolder.widgetPreview,
                appWidgetManager, appWidgetHost,
                appWidgetProviderInfo, appWidgetId)

        viewHolder.widgetLabel.setText(if (adapterItems[position].addedWidgetRecovery) adapterItems[position].widgetLabel + " " + "\uD83D\uDD04" else adapterItems[position].widgetLabel)
        viewHolder.widgetLabel.setTextColor(if (PublicVariable.themeLightDark) widgetConfigurationsActivity.getColor(R.color.dark) else widgetConfigurationsActivity.getColor(R.color.light))

        val drawFloatTheWidget = widgetConfigurationsActivity.getDrawable(R.drawable.draw_open) as LayerDrawable?
        val backFloatTheWidget = drawFloatTheWidget?.findDrawableByLayerId(R.id.backgroundTemporary)
        backFloatTheWidget?.setTint(functionsClass.extractDominantColor(functionsClass.applicationIcon(adapterItems[position].packageName)))

        viewHolder.floatTheWidget.setImageDrawable(drawFloatTheWidget)
        viewHolder.floatTheWidget.setOnClickListener {

            functionsClass.runUnlimitedWidgetService(adapterItems[position].appWidgetId, adapterItems[position].widgetLabel)
        }

        viewHolder.floatTheWidget.setOnLongClickListener { view ->
            val appWidgetProviderInfoLongClick = adapterItems[position].appWidgetProviderInfo

            functionsClassWidgets.popupOptionWidget(widgetConfigurationsActivity, widgetConfigurationsActivity, view,
                    adapterItems[position].packageName,
                    adapterItems[position].classNameProviderWidget,
                    adapterItems[position].appWidgetId,
                    adapterItems[position].widgetLabel,
                    if (appWidgetProviderInfoLongClick.loadPreviewImage(widgetConfigurationsActivity, DisplayMetrics.DENSITY_LOW) != null) {
                        appWidgetProviderInfoLongClick.loadPreviewImage(widgetConfigurationsActivity, DisplayMetrics.DENSITY_HIGH)
                    } else {
                        appWidgetProviderInfoLongClick.loadIcon(widgetConfigurationsActivity, DisplayMetrics.DENSITY_HIGH)
                    },
                    adapterItems[position].addedWidgetRecovery,
                    functionsClass)

            functionsClass.doVibrate(77)

            false
        }

        viewHolder.widgetLabel.setOnEditorActionListener { textView, actionId, keyEvent ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                if (textView.length() > 0) {

                    CoroutineScope(Dispatchers.IO).launch {

                        val widgetDataInterface = Room.databaseBuilder(widgetConfigurationsActivity, WidgetDataInterface::class.java, PublicVariable.WIDGET_DATA_DATABASE_NAME)
                                .fallbackToDestructiveMigration()
                                .addCallback(object : RoomDatabase.Callback() {
                                    override fun onCreate(supportSQLiteDatabase: SupportSQLiteDatabase) {
                                        super.onCreate(supportSQLiteDatabase)
                                    }

                                    override fun onOpen(supportSQLiteDatabase: SupportSQLiteDatabase) {
                                        super.onOpen(supportSQLiteDatabase)
                                    }
                                })
                                .build()
                        widgetDataInterface.initDataAccessObject().updateWidgetLabelByWidgetIdSuspend(adapterItems[position].appWidgetId, textView.text.toString().replace("\uD83D\uDD04", ""))
                        widgetDataInterface.close()

                        withContext(Dispatchers.Main) {
                            //Edit Folder Name
                            SearchEngine.clearSearchDataToForceReload()

                            widgetConfigurationsActivity.forceLoadConfiguredWidgets()
                        }
                    }
                }
            }
            true
        }
    }

    override fun getItemCount(): Int {

        return adapterItems.size
    }

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        var widgetItem: RelativeLayout = view.findViewById<RelativeLayout>(R.id.widgetItem) as RelativeLayout
        var widgetPreview: RelativeLayout = view.findViewById<RelativeLayout>(R.id.widgetPreview) as RelativeLayout
        var floatTheWidget: ImageView = view.findViewById<ImageView>(R.id.floatTheWidget) as ImageView
        var widgetLabel: TextInputEditText = view.findViewById<TextInputEditText>(R.id.widgetLabel) as TextInputEditText
    }
}